crates/admin/src/cluster_controller/scheduler.rs

```diff
@@ -174,6 +436,271 @@ impl<T: TransportConnect> Scheduler<T> {
+                info!("Successfully transitioned from partition processor configuration {} to {}", current_version, next_version);
+                let (_, _, current, next) = epoch_metadata.into_inner();
+                Ok(PartitionConfigurationUpdate {
+                    current,
+                    next,
+                })
+            }
+            Err(ReadModifyWriteError::FailedOperation(concurrent_update)) => {
+                Ok(concurrent_update)
+            }
+            Err(ReadModifyWriteError::ReadWrite(err)) => {
+                Err(err.into())
+            }
+        }
+    }
+
+    /// Checks whether the given partition requires reconfiguration. A partition requires
+    /// reconfiguration in the following cases:
+    ///
+    /// * next contains a replica that is no longer a valid worker candidate.
+    /// * current contains a replica that is no longer a valid worker candidate, and there is no
+    ///   ongoing reconfiguration happening already
+    ///
+    /// In this case, the method returns true, otherwise false.
+    fn requires_reconfiguration(
+        partition_state: &PartitionState,
+        nodes_config: &NodesConfiguration,
+    ) -> bool {
+        let current_requires_reconfiguration =
+            partition_state.current.replica_set().iter().any(|replica| {
+                !nodes_config
+                    .find_node_by_id(*replica)
+                    .map(|node_config| worker_candidate_filter(*replica, node_config))
+                    .unwrap_or_default()
+            });
+
+        let next_requires_reconfiguration = partition_state.next.as_ref().map(|next| {
+            next.replica_set().iter().any(|replica| {
+                !nodes_config
+                    .find_node_by_id(*replica)
+                    .map(|node_config| worker_candidate_filter(*replica, node_config))
+                    .unwrap_or_default()
+            })
+        });
+
+        let ongoing_reconfiguration = next_requires_reconfiguration.is_some();
+        let next_requires_reconfiguration = next_requires_reconfiguration.unwrap_or(false);
+
+        (current_requires_reconfiguration && !ongoing_reconfiguration)
+            || next_requires_reconfiguration
+    }
+
+    fn choose_partition_configuration(
+        partition_id: PartitionId,
+        nodes_config: &NodesConfiguration,
+        partition_replication: ReplicationProperty,
+        preferred_nodes: NodeSet,
+    ) -> Option<PartitionConfiguration> {
+        let options =
+            SelectorOptions::new(u64::from(partition_id)).with_preferred_nodes(preferred_nodes);
+
+        BalancedSpreadSelector::select(
+            nodes_config,
+            &partition_replication,
+            worker_candidate_filter,
+            &options,
+        )
+        .map(|replica_set| {
+            PartitionConfiguration::new(partition_replication, replica_set, HashMap::default())
+        })
+        .inspect_err(|err| {
+            debug!(
```

Comment:
Potentially we can improve error reporting here but let's see how far this one takes us.